### DISCUSSION TITLE
javachecker/windows-x64/javachecker.exe of MirthAdminLauncher causes extrem delay on servers without internet access

### DISCUSSION THREAD

**Discussion by NoraTheExplorer** [nextgenhealthcare/connect#4944] [Q&A]
Hi,

I occasionally stumble across servers without internet connection. When ever I install MirthConnect 3.10 on them, the Launcher takes forever to load the login windows of the MirthConnect Administrator.
With forever I mean about 217 to 220 seconds! This is pretty much reproduceable.

On servers without a network connection at all or with an internet connection it takes only about 27 seconds.

I can see the timing in the launcher.log:
`INFO  2022-01-07 16:48:25,205 [Codebase Downloader Thread] com.mirth.connect.client.launcher.MirthClientLauncher: Time to download/verify codebase: 217.6946212 seconds`

That's a little long, right?

I went through the logfile and found these entries:
```
...
INFO 2022-01-07 16:26:16,791 [pool-2-thread-58] com.mirth.connect.client.launcher.f: Signature verification failed for file openjfx.jar. Downloading latest version...
INFO 2022-01-07 16:26:16,791 [pool-2-thread-22] com.mirth.connect.client.launcher.f: Signature verification failed for file hapi-1.1.jar. Downloading latest version...
...
```

I deleted the cache subfolder and tried again.
The javachecker.exe rebuild the cache\3.10.1 directory and I saw entries in the log like this:
`INFO  2022-01-07 17:01:00,018 [pool-2-thread-22] com.mirth.connect.client.launcher.f: Verified new JAR: C:\Program Files\Vyaire\mirth\MirthAdminLauncher\cache\3.10.1\core\hapi-1.1.jar`

So I assume the javachecker.exe is running into fixed timeouts because the timing is pretty reliable.

Did anyone made the same experience?

**Comment by pacmano1** [comment]
It has been discussed a few times here in github and on the forums.   It is attempting to validate jars signatures as you noted.   I think a workaround mentioned was editing the local hosts file and adding an entry for the endpoint mirth is using to validate the jar.   Search here on GH or the forums you will see related topics.

**Comment by NoraTheExplorer** [comment]
Thank you, I've seen the discussions #4807 and #4866 but none of the suggested workarounds helped.
I'm using Windows Server 2019, connected to one domain network and three private VLAN.
There is no internet access at all.

I edited my hosts file according to this:
```
# localhost name resolution is handled within DNS itself.
127.0.0.1       localhost
::1             localhost
127.0.0.1       ocsp.digicert.com
93.184.220.29   localhost
72.21.91.29     localhost
```

So every request that goes out to the OCSP server should be redirected to my localhost and immediately get an error. 
But this does not happen. It's just extremely slow.

I also tried to disable certain feature of java to not check online for new certificates like suggested in #4679 but with no impact at all.

The whole thing is not a show stopper but extremely annoying.

**Reply by pacmano1** [reply]
Those edits are on your workstation not the server running Mirth?  Assuming they are different of course.

**Reply by pacmano1** [reply]
I'll futz with this later today and see if I can replicate the behavior.  But please do confirm what host you made your edits.

**Reply by NoraTheExplorer** [reply]
Thanks for the fast reply. I called the Launcher directly on the server and also made the changes on the server itself. 
There is no second system involved yet.

**Comment by EDIflyer** [comment]
Just to say I've noticed the same issue on our server today - 498s start time in the log (although in reality it seems to be more like at least 10 minutes), same message re signature verification failure. Tried the hosts change but it didn't seem to speed anything up. Internet access very restricted on this server (running the launcher on the same server as the Mirth service too).

**Reply by jonbartels** [reply]
Try taking a thread dump.

**Reply by EDIflyer** [reply]
@jonbartels thanks - sorry took a while to sort out how to do that as running JRE rather than JDK on the server involved. Have managed it now though - see attached. Any advice much appreciated - when MCAL finally loads then it works OK but it takes 10+ min to load so a bit of a pain if it is closed by mistake, server has been rebooted, etc.
[threaddump.txt](https://github.com/nextgenhealthcare/connect/files/8837433/threaddump.txt)

**Reply by jonbartels** [reply]
Sorry for the short reply earlier.

What we can see in that thread dump with a simple text search is:

* No threads are `BLOCKED`
* A few threads are `WAITING` but these look like ordinary pauses while other tasks complete
* Many threads are `RUNNABLE`

Of those `RUNNABLE` threads we see some common stack traces. There are over 100 of these:
```
"pool-2-thread-156" #208 prio=5 os_prio=0 tid=0x0000000022efa800 nid=0x1cc0 runnable [0x00000000303fd000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(Unknown Source)
	at java.net.SocketInputStream.read(Unknown Source)
	at java.net.SocketInputStream.read(Unknown Source)
	at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:137)
	at org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:153)
	at org.apache.http.impl.io.SessionInputBufferImpl.readLine(SessionInputBufferImpl.java:280)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:138)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:56)
	at org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:259)
	at org.apache.http.impl.DefaultBHttpClientConnection.receiveResponseHeader(DefaultBHttpClientConnection.java:163)
	at org.apache.http.protocol.HttpRequestExecutor.doReceiveResponse(HttpRequestExecutor.java:273)
	at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:125)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:272)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
	at com.mirth.connect.client.launcher.c.b(SourceFile:92)
```

So MCAL is doing something with an HTTP call. It has over 100 concurrent HTTP calls when it pauses. 

The stack trace gets into obfusucated MCAL source code. Only NextGen can de-obfuscate that and know what call is being made.

One thing you might be able to do when MCAL pauses is to look at `netstat`. What server(s) is the MCAL process trying to talk to?

**Reply by jonbartels** [reply]
Oh let's also rule out the simple stuff first - Are you using the latest version of MCAL?

**Reply by EDIflyer** [reply]
Thanks @jonbartels.  Re netstat - there are lots of connections already open on the server but I see the -b option shows the requesting application so will have a go looking at that.

Re MCAL version - I can't see any about info in the launcher that lists the version number of it? I'll redownload the latest version from https://www.nextgen.com/products-and-services/mirth-connect-downloads just in case though.

**Reply by EDIflyer** [reply]
OK interesting - the netstat -b is flooded with requests to 172.64.155.188 and 104.18.32.68 from launcher.exe.  From what I can see both of these are part of CloudFlare? e.g., https://whois.arin.net/rest/net/NET-172-64-0-0-1/pft?s=172.64.155.188  Not sure if that helps or not but certainly seems to be where all the connections are sitting waiting.

It's also happening over hundreds of ports - 52089 to 52455 to what I can see at a quick glance

PS the above was after ensuring MCAL 1.2.0 64-bit (ie latest version) installed - there was no change in behaviour with it.

**Reply by pacmano1** [reply]
Isn't this the same issue already discussed a few times here on Github about validating jars? Last one I remember was related to firewalls blocking the requests because they looked suspicious.

**Reply by jonbartels** [reply]
Hmm. Lets think about the HTTP traffic MCAL might generate. It's closed source so we have to guess:

* Fetching the MC JNLP file -> This is one thread getting a tiny JNLP XML object
* Fetching the MC JARs to launch the client -> This is 20-30 JARs for the MC client, plugins, and dependencies. This would all be HTTP traffic to the MC server itself
* OCSP/CRL checks -> Each of the JARs will be signed then each certificate in the chain would have an HTTP call made to get the OCSP or CRL listings. If each of the 20-30 JARs has a cert chain that is 3 certificates and MCAL does a check for each one w/o caching or coordination, that would add up to a lot of connections

@EDIflyer - I'm not enough of a sysadmin to know if its possible to block the IPs or forcibly terminate the connections. If you take a packet capture do you see traffic moving over those connections to the IPs you found or just one-way attempts to connect from MCAL to Cloudflare?

Is there a good general strategy that would cause blocked connections to just outright terminate instead of hanging? It might fail but at least it'd fail faster if the connection were terminated.

@twest-mirthconnect - I am tagging you because MCAL is closed source. Could  you please help us get 15 minutes from a NextGen developer resource? If someone from NG can look at thread dump from EDIFlyer and  NoraTheExplorer and tell us what the obfuscated code is really doing at `com.mirth.connect.client.launcher.c.b(SourceFile:92)` that would help identify workarounds for users in restrictive environments. 

This could be as simple as a missing or default timeout. It could also be as complicated as needing a cache/synchronization for OCSP checks.

**Reply by jonbartels** [reply]
@EDIflyer  - I just had an idea. Both MC and MCAL would use the Apache Commons HTTP libraries. I know those have a "wire-level" logging option. What it does is literally spit out all of the HTTP calls that the Apache library is making. It's super noisy but it spits out the headers, URLs, response codes, etc. 

It'll be a while but I'll try to dig out the properties/loggers for that and find out what version of Apache MCAL is using. 

That should help your issue in that if we log out every HTTP call MCAL tries to make, then you can adjust your environment to allow or block or rapidly terminate those calls as needed.

**Reply by EDIflyer** [reply]
@jonbartels if there's an easy way to try and obtain this info and it would be helpful then happy to try and provide. I suspect there'll be some limitations what what environmental adjustments are possible.  I suspect @pacmano1 may be onto something when it comes to firewalls blocking potentially suspicious requests. Given the number of connections being initiated I can see why it would be flagged. It also used to work absolutely fine for us previously - something changed in the last few weeks (we're not on the system daily) and I would be suspicious it's something network-level that is now blocking these connections?

**Reply by jonbartels** [reply]
I'm working through the idea of logging the HTTP calls. Sorry for the spam in GitHub but I hope that by working in the open I can help others and collaborate on the problem. It's a fun one!

Find out where the MC log properties are:
```
/Applications/Mirth Connect Administrator Launcher.app/Contents/java/app
java -Dlog4j.debug=true -jar mirth-client-launcher.jar 
log4j: Trying to find [log4j.xml] using context classloader sun.misc.Launcher$AppClassLoader@14ae5a5.
log4j: Trying to find [log4j.xml] using sun.misc.Launcher$AppClassLoader@14ae5a5 class loader.
log4j: Trying to find [log4j.xml] using ClassLoader.getSystemResource().
log4j: Trying to find [log4j.properties] using context classloader sun.misc.Launcher$AppClassLoader@14ae5a5.
log4j: Using URL [jar:file:/Applications/Mirth%20Connect%20Administrator%20Launcher.app/Contents/java/app/mirth-client-launcher.jar!/log4j.properties] for automatic log4j configuration.
log4j: Reading configuration from URL jar:file:/Applications/Mirth%20Connect%20Administrator%20Launcher.app/Contents/java/app/mirth-client-launcher.jar!/log4j.properties
.. logger details...
```

So the log properties are embedded in the JAR. 

OK that wasn't too hard to set up - Just run MCAL with 
```
 java -Dlog4j.debug=true -Dorg.apache.commons.logging.simplelog.showdatetime=true  -Dorg.apache.commons.logging.simplelog.log.org.apache.http=DEBUG -Dorg.apache.commons.logging.simplelog.log.org.apache.http.wire=ERROR -Dorg.apache.commons.logging.Log=org.apache.commons.logging.impl.SimpleLog -jar mirth-client-launcher.jar 
 ```
 
This will then spit out output like this *to the console*:
```
INFO  2022-06-07 12:33:27,793 [pool-2-thread-28] com.mirth.connect.client.launcher.f: Using cached JAR: /Applications/Mirth Connect Administrator Launcher.app/Contents/java/app/cache/3.9.1/core/jcifs-ng-2.1.3.jar
2022/06/07 12:33:27:836 EDT [DEBUG] DefaultHttpClientConnectionOperator - Connection established 192.168.113.3:57178<->72.21.91.29:80
2022/06/07 12:33:27:836 EDT [DEBUG] DefaultManagedHttpClientConnection - http-outgoing-134: set socket timeout to 60000
2022/06/07 12:33:27:836 EDT [DEBUG] MainClientExec - Executing request POST / HTTP/1.1
2022/06/07 12:33:27:836 EDT [DEBUG] MainClientExec - Target auth state: UNCHALLENGED
2022/06/07 12:33:27:836 EDT [DEBUG] MainClientExec - Proxy auth state: UNCHALLENGED
2022/06/07 12:33:27:836 EDT [DEBUG] headers - http-outgoing-134 >> POST / HTTP/1.1
2022/06/07 12:33:27:836 EDT [DEBUG] headers - http-outgoing-134 >> Content-Type: application/ocsp-request
2022/06/07 12:33:27:836 EDT [DEBUG] headers - http-outgoing-134 >> Accept: application/ocsp-response
2022/06/07 12:33:27:836 EDT [DEBUG] headers - http-outgoing-134 >> Content-Length: 83
2022/06/07 12:33:27:836 EDT [DEBUG] headers - http-outgoing-134 >> Host: ocsp.digicert.com
2022/06/07 12:33:27:836 EDT [DEBUG] headers - http-outgoing-134 >> Connection: Keep-Alive
2022/06/07 12:33:27:836 EDT [DEBUG] headers - http-outgoing-134 >> User-Agent: Apache-HttpClient/4.5.8 (Java/1.8.0_161)
2022/06/07 12:33:27:836 EDT [DEBUG] headers - http-outgoing-134 >> Accept-Encoding: gzip,deflate
2022/06/07 12:33:27:864 EDT [DEBUG] headers - http-outgoing-134 << HTTP/1.1 200 OK
2022/06/07 12:33:27:864 EDT [DEBUG] headers - http-outgoing-134 << Accept-Ranges: bytes
2022/06/07 12:33:27:864 EDT [DEBUG] headers - http-outgoing-134 << Age: 938
2022/06/07 12:33:27:864 EDT [DEBUG] headers - http-outgoing-134 << Cache-Control: 'max-age=158059'
2022/06/07 12:33:27:864 EDT [DEBUG] headers - http-outgoing-134 << Content-Type: application/ocsp-response
2022/06/07 12:33:27:864 EDT [DEBUG] headers - http-outgoing-134 << Date: Tue, 07 Jun 2022 16:33:27 GMT
2022/06/07 12:33:27:864 EDT [DEBUG] headers - http-outgoing-134 << Last-Modified: Tue, 07 Jun 2022 16:17:49 GMT
2022/06/07 12:33:27:864 EDT [DEBUG] headers - http-outgoing-134 << Server: ECS (cha/81BD)
2022/06/07 12:33:27:864 EDT [DEBUG] headers - http-outgoing-134 << X-Cache: HIT
2022/06/07 12:33:27:864 EDT [DEBUG] headers - http-outgoing-134 << Content-Length: 471
2022/06/07 12:33:27:864 EDT [DEBUG] MainClientExec - Connection can be kept alive indefinitely
2022/06/07 12:33:27:865 EDT [DEBUG] BasicHttpClientConnectionManager - Releasing connection 192.168.113.3:57178<->72.21.91.29:80
2022/06/07 12:33:27:865 EDT [DEBUG] DefaultManagedHttpClientConnection - http-outgoing-134: set socket timeout to 0
2022/06/07 12:33:27:865 EDT [DEBUG] BasicHttpClientConnectionManager - Connection can be kept alive indefinitely
2022/06/07 12:33:27:865 EDT [DEBUG] BasicHttpClientConnectionManager - Closing connection
2022/06/07 12:33:27:865 EDT [DEBUG] DefaultManagedHttpClientConnection - http-outgoing-134: Close connection
2022/06/07 12:33:27:866 EDT [DEBUG] RequestAddCookies - CookieSpec selected: default
2022/06/07 12:33:27:866 EDT [DEBUG] RequestAuthCache - Auth cache not set in the context
2022/06/07 12:33:27:866 EDT [DEBUG] BasicHttpClientConnectionManager - Get connection for route {}->http://ocsp.digicert.com:80
2022/06/07 12:33:27:866 EDT [DEBUG] DefaultManagedHttpClientConnection - http-outgoing-625: set socket timeout to 60000
2022/06/07 12:33:27:866 EDT [DEBUG] MainClientExec - Opening connection {}->http://ocsp.digicert.com:80
2022/06/07 12:33:27:867 EDT [DEBUG] DefaultHttpClientConnectionOperator - Connecting to ocsp.digicert.com/72.21.91.29:80
```

@EDIflyer - You should try running MCAL with the HTTP logging options. Capture that output to a text file and you should be able to then trace through what HTTP calls it is *trying* to make in *your* environment and see what fails.

Also look at `log/launcher.log` and see if it captured any HTTP or timeout related exceptions.

**Reply by jonbartels** [reply]
@EDIflyer - Looks like we've got some time/attention from NextGen staff.

Any luck setting up the HTTP wire logging to see what MCAL is trying to talk to?

**Reply by EDIflyer** [reply]
> @EDIflyer - Looks like we've got some time/attention from NextGen staff.
> 
> Any luck setting up the HTTP wire logging to see what MCAL is trying to talk to?

Sorry I'm not quite clear what is required re HTTP wire logging? 'Was is just running MCAL with the options you mentioned?
```
 java -Dlog4j.debug=true -Dorg.apache.commons.logging.simplelog.showdatetime=true  -Dorg.apache.commons.logging.simplelog.log.org.apache.http=DEBUG -Dorg.apache.commons.logging.simplelog.log.org.apache.http.wire=ERROR -Dorg.apache.commons.logging.Log=org.apache.commons.logging.impl.SimpleLog -jar mirth-client-launcher.jar 
```

**Reply by jonbartels** [reply]
Yep. Those options will make the HTTP library that MCAL uses dump a lot of information about what HTTP calls it is trying to make. That should show what calls are being blocked and help figure out how to make them fail faster so your startup doesn't take so long.

**Reply by EDIflyer** [reply]
@jonbartels OK have done that (had to do `2>&1` to pipe to a file as most was just going to the console!), have a nice chunky 4MB log file now (see attached)
[debuglogging.txt](https://github.com/nextgenhealthcare/connect/files/8871449/debuglogging.txt)

**Reply by jonbartels** [reply]
That log shows over TWO THOUSAND outgoing HTTP requests. `grep -o "http-outgoing-\d\+" debuglogging.txt | sort -u | cut -d "-" -f 3 | sort -n`

The unique URLs its hitting are:
```
grep "Opening connection" debuglogging.txt | cut -d " " -f 9 | sort | uniq -c
 348 {}->http://crl.comodo.net:80
 696 {}->http://crl.comodoca.com:80
 348 {}->http://crl.sectigo.com:80
 348 {}->http://crl.usertrust.com:80
   1 {}->http://localhost:8080
 174 {}->http://ocsp.comodoca.com:80
 174 {}->http://ocsp.sectigo.com:80
 174 {}->http://ocsp.usertrust.com:80
 ```
 
 Then we can do a DNS lookup on those to correlate them to the IPs that are apparently blocked:
 ```
 grep "Opening connection" debuglogging.txt | cut -d " " -f 9 | sort | uniq -c | awk -F[/:] '{print $4}' | sort -u | xargs -L1 -t dig +short
dig +short crl.comodo.net
dig +short crl.comodoca.com
crl.comodoca.com.cdn.cloudflare.net.
104.18.32.68
172.64.155.188
dig +short crl.sectigo.com
crl.comodoca.com.cdn.cloudflare.net.
172.64.155.188
104.18.32.68
dig +short crl.usertrust.com
crl.comodoca.com.cdn.cloudflare.net.
104.18.32.68
172.64.155.188
dig +short localhost
127.0.0.1
dig +short ocsp.comodoca.com
ocsp.comodoca.com.cdn.cloudflare.net.
172.64.155.188
104.18.32.68
dig +short ocsp.sectigo.com
ocsp.comodoca.com.cdn.cloudflare.net.
172.64.155.188
104.18.32.68
dig +short ocsp.usertrust.com
ocsp.comodoca.com.cdn.cloudflare.net.
172.64.155.188
104.18.32.68 
 ```
 
And there you go - comodo is using cloudflare for OCSP/CRL.

Your server startup takes forever because MCAL makes TWO THOUSAND queries to a URL that is blocked by your network. Even if its blocked as fast and polite as possible that's going to take time and contribute to `INFO  2022-06-09 15:33:56,166 [Codebase Downloader Thread] com.mirth.connect.client.launcher.MirthClientLauncher: Time to download/verify codebase: 493.772518638 seconds`

**Reply by EDIflyer** [reply]
Thanks @jonbartels - no wonder the firewalls flagged the IP! Thanks for doing this investigative work 👍

**Reply by jonbartels** [reply]
I ran the same analysis on MCAL for one my my instances:
```
grep "Opening connection" jonb-mcal-http.log | cut -d " " -f 9 | sort | uniq -c
   1 {}->http://10.31.16.9:8080
 164 {}->http://crl3.digicert.com:80
  31 {}->http://crl4.digicert.com:80
 370 {}->http://ocsp.digicert.com:80
 ```
 
 That's far fewer connections than EDIFlyer had but mine also aren't blocked. I know that the digicert calls are because of the Zen extensions, they're signed by a digicert CA. 100s of HTTP calls for what can't be more than 3 or 4 certificates in a chain is a lot.

**Reply by jonbartels** [reply]
@EDIflyer what non-NextGen extensions or MC versions are you running? I ask because the root CA from NextGen and Mirth has been Digicert going back as far as MC 3.3.x. What are you running that's signed by Comodo?

This is also notable because surely there should be a call to digicert in your log output to validate NextGens own certificates.

**Reply by EDIflyer** [reply]
@jonbartels Pdfbox and fontbox.... The only jar we add is fontbox... We use the pdfbox version that comes with mirth.

**Reply by jonbartels** [reply]
To wrap this up - Peter from Nextgen confirmed our findings. https://github.com/nextgenhealthcare/connect/discussions/4944#discussioncomment-2942494 

In the short term it looks like you need to negotiate with your network team to tolerate the connections.

**Reply by EDIflyer** [reply]
Thanks for your help, @jonbartels - at least the issue has been confirmed. Once it's up and running things are OK so we might not bother trying to negotiate things for now but handy for us to now have a full list of URLs that need to be permitted from your post above (https://github.com/nextgenhealthcare/connect/discussions/4944#discussioncomment-2914912)

**Comment by siddharthsharma9258** [comment]
The expectation is that mirth should be able to run without internet connection?

**Reply by jonbartels** [reply]
OP and the recent comment from @EDIflyer are referring to the "Mirth Connect Launcher" (MCAL) needing internet access to run. 

MCAL needs internet access to do OCSP checks against the MC client and plugin signing certificates.

**Comment by twest-mirthconnect** [comment]
@jonbartels - Yes, you can have a little time though we must cap it at 2 hrs total time. We are under the gun with CURES certification work and have a release in a few weeks.

Feel free to reach out to @pladesma this week.

**Reply by jonbartels** [reply]
Oh cool! Thanks for wrangling that Travis. 

@pladesma I hope this is quick. Since MCAL is closed source I need a developer to confirm a few things to help NoraTheExplorer, EDIFlyer, and others who have reported similar issues.

1. For the stack traces and logs that show a failure at `com.mirth.connect.client.launcher.c.b(SourceFile:92)`, what is that code doing? Is this actually the OCSP check?
2. Is there a logger or configurable option that would let MCAL skip or fail-fast with OCSP checks for restrictive environments that block the outbound OCSP queries?
3. Per https://github.com/nextgenhealthcare/connect/issues/4866 does MCAL do any caching, synchronization, etc. to reduce the number of external OCSP calls made? From what I can see it is acting like its doing one OCSP check per certificate per signature validation, even if the JARs its validating have the same certificate chain. Can you confirm that my external observations are correct?

If a phone call or Slack would help, please DM me.

I do not have a personal stake in this issue but several users have filed good questions about MCAL and this OCSP behavior. I want to advocate for them. MCAL is pretty handy and helping it run better is valuable.

**Reply by pladesma** [reply]
1. This line of code executes an HTTP request to get a Certificate Revocation List (CRL). An error at this line indicates a failure to download the CRL.
2. No, currently there aren't any options to skip or fail-fast.
3. Your observations are correct. There's isn't any caching or anything else to reduce the number of OCSP calls.

That may not be too helpful, but your questions point to some good improvements we can make.

**Reply by jonbartels** [reply]
Thanks @pladesma . That answers my questions. I appreciate your time looking at closed-source code to confirm what EDIFlyer and I found.

**Comment by pacmano1** [comment]
If you configure your firewall to reject (not block) ths connection everything should run fast.

The difference between block and reject is that with reject, a packet (TCP RST or ICMP port unreachable for UDP) is returned to the sender, whereas with block the packet is dropped silently. In either case, the original packet is discarded.

So with ufw in linux you can do:   `ufw reject...`

**Reply by EDIflyer** [reply]
@pacmano1 so would this be something that could be set in the hosts file? The server is running Windows Sever 2016, not Linux I'm afaid.

**Reply by pacmano1** [reply]
You will need to do that at the network level, I don't think windows firewall supports a RST.

**Reply by EDIflyer** [reply]
> You will need to do that at the network level, I don't think windows firewall supports a RST.

Ah OK - that is likely much trickier to try and sort given the size of the organisation we're dealing with. If this is the issue then we might be better trying to ask them to whitelist the connection (as I say it worked previously and only recently seems to have been blocked, presumably due to the automated systems flagging the number of connections being opened). I guess the question is do they just use these two Cloudflare IPs or are there other ones that they might use too?
